packages/framework/aqueduct/src/data-object-factories/pureDataObjectFactory.ts

```diff
@@ -101,26 +106,135 @@ class FluidDataObjectFactory {
+        // as that results in creation of DDSs, before it gets attached, providing atomic
+        // guarantee of creation.
+        // WARNING: we can't do the same (yet) for already existing PureDataObject!
+        // This will result in deadlock, as it tries to resolve internal handles, but any
+        // handle resolution goes through root (container runtime), which can't route it back
+        // to this data store, as it's still not initialized and not known to container runtime yet.
+        // In the future, we should address it by using relative paths for handles and be able to resolve
+        // local DDSs while data store is not fully initialized.
+        if (!instance.runtime.existing) {
+            await instance.getInstance();
+        }
+
+        return instance;
+    }
+
+    protected constructor(
+        private readonly ctor: new (props: IDataObjectProps<O>) => TObj,
+        readonly context: IFluidDataStoreContext,
+        private readonly sharedObjectRegistry: ISharedObjectRegistry,
+        private readonly optionalProviders: FluidObjectSymbolProvider<O>,
+        private readonly props: S,
+        onDemandInstantiation: boolean)
+    {
+        // base
+        let runtimeCtor = FluidDataStoreRuntime;
+
+        // request mixin in
+        runtimeCtor = requestFluidDataStoreMixin(
+            runtimeCtor,
+            async (request: IRequest) => (await this.getInstance()).request(request));
+
+        // If factory needs to instantiate PureDataObject, then it means it needs to finish
+        // initialization before anything can be done with runtime. For example, DataObject needs
+        // to provide additional blobs to summary that are not part of DDS.
+        // Please note that ideal solution is for code above to await instanceP, but given that we can't
+        // do it yet for existing data store, we have to do it only when summary is called, i.e. after
+        // data store runtime is fully initialized (from POV of container runtime)
+        if (!onDemandInstantiation) {
+            runtimeCtor = summaryWaitFluidDataStoreMixin(
+                runtimeCtor,
+                async () => {await this.getInstance(); });
+        }
+
+        // Create a new runtime for our data store
+        // The runtime is what Fluid uses to create DDS' and route to your data store
+        this.runtime = new runtimeCtor(
+            context,
+            this.sharedObjectRegistry,
+        );
+
+        // Create object right away if required.
+        // This allows object to register various callbacks with runtime before runtime
+        // becomes globally available. But it's not full initialization - constructor can't
+        // access DDSs or other services of runtime as objects are not fully initialized.
+        // In order to use object, we need to go through full initialization by calling getInstance().
+        if (!onDemandInstantiation) {
```

Comment:
Ideally, this we always create DataObject, but not initialize it until needed (and remove onDemandInstantiation).
I.e. all expensive work happens  in initializingFromExisting / initializingFirstTime / hasInitialized methods, but ctor is cheap.
@leeviana , @aboktor, do you think this is reasonable to change, or it's part of life?
AtMentionsCollection seems like the biggest offender here. All other examples seems to have trivial ctors.

Reply:
Yeah, I think constructor is generally pretty cheap, although there are at least 4 places in component lifecycle that I'm aware of that expensive work can happen and it's not just initialize:

- in the factory create (for getting services, setting them up etc. because you have access to context/runtime, although I think now most of this can happen in initialize), this includes constructor cost
- in initialize
- in request (this is important/key for the scriptor path because until request happens, scriptor does not know extra context things like maybe start/end marker. Thus scriptor can't initialize it's model (with virtualization etc.) until the request
- in render (just because render is expensive)

Reply:
So, do I hear that I should move in this direction and fix AtMentionsCollecitons' ctor to be cheap?

Reply:
Yes, AtMentionsCollections ctor logic can most likely move to initialize :)

Reply:
Switching to always creating DataObjects, but finish initialization only on demand
I'll raise PR in parallel for At-mention collection.